### PR TITLE
(experimental) Calibrate networkpolicy test aggressiveness

### LINKED ISSUE #103802: reduce resource requirements of networkpolicy tests by 1/3
#### Which jobs are flaking:

The `NetPol` tests can take a long time to start since they require 9 pods.    This can result in failures like: 

```fail [k8s.io/kubernetes@v1.21.1/test/e2e/network/netpol/network_policy.go:1208]: unable to initialize resources```

Now, these tests were designed to fail fast - i.e. if you cant make x pods, dont bother testing.  however... We can reduce the odds that such a scenario happens i think by 

- reducing the cost of making x pods  (1) below
- reducing the number X (2 or 3)

Fix ideas below:

0) Potential super-easiest fix: request 50 millicored per `agnhost` pod.  Since the pods literally do nothing other then respond to one or two web requests... maybe thats enough to make the scheduler 50% more likely to create the pods quickly enough.

1) Potential easiest fix: reduce pod footprint by making 
```
func (p *Pod) KubePod() *v1.Pod {
```
configurable to take a protocol 
```
func (p *Pod) KubePod(protocols []v1.Protocol)
```
This would potentially reduce resource requirements by 66%

2)  have each test specify positive / negative control set and make model smart enough to only create required pods - and poller smart enough to poll required pods.  This makes the tests arbitrarily resource efficient - but then re-adds the original problem of now, any failures we get are going to require fine grained understanding of the test plan and lots of manual curation of exact testing requirements.  This is imo unsustainable. 

3) Have each test specify a "region" of the matrix.  i.e. "requiredNamespaces=x" would only poll this matrix and thus only require 3 pods instead of 9.... 

``` x/a x/b x/c
x/a
x/b
x/c
```

note that (3) and (1) could be combined.

...  given that the flake rate is 1/10 right now, i think reducing the resources by 66% (which you get for free w/ (1)) would be good enough as a first pass.

#### Testgrid link:

https://k8s-testgrid.appspot.com/sig-network-gce#network-policies,%20google-gce

### PR DESCRIPTION
**Pull request by jayunit100**
fixes #103802

A quick experiment to see if these tests can be less aggressive in terms of resource usage and thus faster in completion  time, and less associated w correlated flakes in the signet test suite.

It seems like the networkpolicy flake rate is increasing lately (not sure why... maybe our overall sig-network test suite is getting larger or we've added new networkpolicy tests...)

### CODE REVIEW

**Comment by jayunit100** [comment]
last failure was 
```


3/6687 Tests Failed. | expand_less
-- | --
Kubernetes e2e suite: [sig-network] NetworkPolicyLegacy [LinuxOnly] NetworkPolicy between server and client should allow ingress access from updated namespace [Feature:NetworkPolicy] expand_more3m0s | Kubernetes e2e suite: [sig-network] NetworkPolicyLegacy [LinuxOnly] NetworkPolicy between server and client should allow ingress access from updated namespace [Feature:NetworkPolicy] expand_more | 3m0s
Kubernetes e2e suite: [sig-network] NetworkPolicyLegacy [LinuxOnly] NetworkPolicy between server and client should allow ingress access from updated namespace [Feature:NetworkPolicy] expand_more | 3m0s
Kubernetes e2e suite: [sig-network] Networking Granular Checks: Services should function for client IP based session affinity: udp [LinuxOnly] expand_more | Kubernetes e2e suite: [sig-network] Networking Granular Checks: Services should function for client IP based session affinity: udp [LinuxOnly] expand_more
Kubernetes e2e suite: [sig-network] Networking Granular Checks: Services should function for client IP based session affinity: udp [LinuxOnly] expand_more


``` 

so, it looks like this fixes any lingering netpol issues, but i just put another one in to reduce the cpu usage by another 1/3, i.e. by checking what protocols are needed. 

expect ci to fail bc i only am enabling TCP this time... will enable others if the results look promising

**Comment by jayunit100** [comment]
*weird* why does this job keep failing ?

File: test/e2e/network/netpol/test_helper.go [hunk 1]
```diff
@@ -30,7 +30,8 @@ import (
 )
 
 const (
-	waitInterval = 1 * time.Second
+	// increased to 5 seconds, from 1, so that the test waits longer to start in case of congestion
```
Comment (jayunit100) [inline test/e2e/network/netpol/test_helper.go:33]: note ~ empirically, we're finding that these new tests run 30% faster, so this isnt actually
slowing anything down.

File: test/e2e/network/netpol/test_helper.go [hunk 2]
```diff
@@ -67,7 +68,7 @@ func UpdatePolicy(k8s *kubeManager, policy *networkingv1.NetworkPolicy, namespac
 
 // waitForHTTPServers waits for all webservers to be up, on all protocols sent in the input,  and then validates them using the same probe logic as the rest of the suite.
 func waitForHTTPServers(k *kubeManager, model *Model) error {
-	const maxTries = 10
+	const maxTries = 20
```
Comment (jayunit100) [inline test/e2e/network/netpol/test_helper.go:71]: see comment above on max tries, this is only needed in case of extreme congestion, but it wont slow things down from what were seeing in measurements.

File: test/e2e/network/netpol/test_helper.go (hunk 1 above)
Comment (jayunit100) [reply test/e2e/network/netpol/test_helper.go:33]: @rajaskakodkar , can you make the call on wether or not we stil need this?

File: test/e2e/network/netpol/test_helper.go (hunk 2 above)
Comment (jayunit100) [reply test/e2e/network/netpol/test_helper.go:71]: @rajaskakodkar , can you make the call on wether or not we stil need this?

File: test/e2e/network/netpol/model.go [hunk 3]
```diff
@@ -325,4 +367,7 @@ func (c *Container) Spec() v1.Container {
 			},
 		},
 	}
+	// reduce cpu usage 10 fold compared to a normal pod...
```
Comment (jayunit100) [inline test/e2e/network/netpol/model.go:370]: lets improve this comment?

File: test/e2e/network/netpol/network_policy.go [hunk 4]
```diff
@@ -123,7 +120,15 @@ var _ = common.SIGDescribe("Netpol", func() {
 
 	ginkgo.Context("NetworkPolicy between server and client", func() {
 		ginkgo.BeforeEach(func() {
-			model = initializeResourcesByFixedNS(f)
+			// These flags control which containers get created in the default model. Create containers with TCP protocol only for tests in this context.
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:123]: s/flags/variables

File: test/e2e/network/netpol/model.go [hunk 5]
```diff
@@ -21,11 +21,20 @@ import (
 	"strings"
 
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/kubernetes/test/e2e/framework"
 	imageutils "k8s.io/kubernetes/test/utils/image"
 )
 
+// getResourceRequirements returns a ResourceRequirements object
```
Comment (jayunit100) [inline test/e2e/network/netpol/model.go:30]: ... add note about how this was borrowed from another place in the code base, or else, see if we can reuse the original code that i copied this from :)

File: test/e2e/network/netpol/test_helper.go (hunk 1 above)
Comment (rajaskakodkar) [reply test/e2e/network/netpol/test_helper.go:33]: I think we can drop this

File: test/e2e/network/netpol/test_helper.go (hunk 2 above)
Comment (rajaskakodkar) [reply test/e2e/network/netpol/test_helper.go:71]: I think we can revert this

**Comment by jayunit100** [comment]
We can close #103802 after this merges

File: test/e2e/network/netpol/model.go (hunk 5 above)
Comment (rajaskakodkar) [reply test/e2e/network/netpol/model.go:31]: i have pushed a commit with a note about where this was borrowed from, I don't think we can import it from the code base. A quick search https://cs.k8s.io/?q=getResourceRequirements&i=nope&files=&excludeFiles=&repos= shows that this function is defined in other parts of codebase the way we have done it

lmk if you think otherwise

File: test/e2e/network/netpol/model.go [hunk 6]
```diff
@@ -38,11 +48,14 @@ type Model struct {
 	Ports          []int32
 	Protocols      []v1.Protocol
 	DNSDomain      string
+	UDP            bool
```
Comment (mattfenwick) [inline test/e2e/network/netpol/model.go:51]: This is already covered by the `Protocols` field -- why not modify that instead?

File: test/e2e/network/netpol/model.go [hunk 7]
```diff
@@ -71,10 +84,24 @@ func NewModel(namespaces []string, podNames []string, ports []int32, protocols [
 			var containers []*Container
 			for _, port := range ports {
 				for _, protocol := range protocols {
-					containers = append(containers, &Container{
```
Comment (mattfenwick) [inline test/e2e/network/netpol/model.go:74]: I don't understand this change -- isn't this just a more verbose way of doing the same thing?

File: test/e2e/network/netpol/network_policy.go [hunk 8]
```diff
@@ -123,7 +120,15 @@ var _ = common.SIGDescribe("Netpol", func() {
 
 	ginkgo.Context("NetworkPolicy between server and client", func() {
 		ginkgo.BeforeEach(func() {
-			model = initializeResourcesByFixedNS(f)
+			// These variables control which containers get created in the default model. Create containers with TCP protocol only for tests in this context.
+			var tcp, udp, sctp = true, false, false
+			for _, component := range ginkgo.CurrentGinkgoTestDescription().ComponentTexts {
+				// Create UDP containers along with TCP for this particular test.
+				if component == "should enforce ingress policy allowing any port traffic to a server on a specific protocol [Feature:NetworkPolicy] [Feature:UDP]" {
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:127]: The information of whether to use a protocol or not should be provided in the test case itself -- otherwise, the tests will be difficult to comprehend.

File: test/e2e/network/netpol/network_policy.go [hunk 9]
```diff
@@ -1163,7 +1168,9 @@ var _ = common.SIGDescribe("Netpol [LinuxOnly]", func() {
 
 	ginkgo.Context("NetworkPolicy between server and client using UDP", func() {
 		ginkgo.BeforeEach(func() {
-			model = initializeResourcesByFixedNS(f)
+			// Create containers with only UDP protocol in the default model used for tests in this context.
+			tcp, udp, sctp := false, true, false
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:1172]: This should pass in a slice of protocols instead of three separate booleans.

**Review by mattfenwick — CHANGES_REQUESTED** [review]
The protocols already exist in the model -- instead of duplicating them, the new logic to control them should integrate with the existing code.

File: test/e2e/network/netpol/model.go [hunk 10]
```diff
@@ -325,4 +368,7 @@ func (c *Container) Spec() v1.Container {
 			},
 		},
 	}
+	// Request 1m of CPU and 50Mi of memory for each of the containers created in the pod and set a limit of 75m and 75Mi
+	container.Resources = getResourceRequirements(getResourceList("1m", "50Mi"), getResourceList("75m", "75Mi"))
```
Comment (knabben) [inline test/e2e/network/netpol/model.go:372]: I would bring them to constants at the top of the model this way.

File: test/e2e/network/netpol/network_policy.go (hunk 8 above)
Comment (rajaskakodkar) [reply test/e2e/network/netpol/network_policy.go:127]: @mattfenwick ya, that's kind of a chicken and egg problem... we need to pass the information about protocols while creating the model and the logic for that lives in `BeforeEach`. If we move the information about protocol in the test case then we will have to move the model logic there. 

Kind of adding 
```
model = initializeResourcesByFixedNS(f, protocols)
```
to each test case

ref: https://github.com/kubernetes/kubernetes/pull/102354#discussion_r668704334

**Comment by rajaskakodkar** [comment]
@mattfenwick @knabben I have addressed your review comments and added a commit for the same, PTAL

cc: @jayunit100

File: test/e2e/network/netpol/model.go (hunk 7 above)
Comment (rajaskakodkar) [reply test/e2e/network/netpol/model.go]: Yes, once we have a slice of protocols passed, we don't need this check

File: test/e2e/network/netpol/model.go [hunk 11]
```diff
@@ -21,11 +21,28 @@ import (
...
 
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/kubernetes/test/e2e/framework"
 	imageutils "k8s.io/kubernetes/test/utils/image"
 )
 
+const (
+	CPURequest    = "1m"
```
Comment (mattfenwick) [inline test/e2e/network/netpol/model.go:31]: Where did these constants come from?  Aren't the containers just running agnhost or something?

I'm surprised they'd need this much CPU and memory, can we collect some data showing how much they use during a test run?

File: test/e2e/network/netpol/model.go [hunk 12]
```diff
@@ -21,11 +21,28 @@ import (
...
+	MemoryRequest = "50Mi"
+	CPULimit      = "75m"
```
Comment (mattfenwick) [inline test/e2e/network/netpol/model.go:33]: Why not use `resource.MustParse` here so you don't have to parse resources for every single container?

File: test/e2e/network/netpol/network_policy.go (hunk 8 above)
Comment (mattfenwick) [reply test/e2e/network/netpol/network_policy.go:127]: This is a readability vs. convenience tradeoff, and currently this PR is giving away readability for a tiny amount of convenience.

I'm against this approach, for the reason I gave above -- this gives tests surprising behavior which is going to confuse developers.

Since protocols are determined by test case, it makes total sense that each test case has to state which protocols it wants.

You may have been misled by the `addSCTPContainers` flag, which originally was only for developer debugging but then at some point looks like it was wrangled in to determining test behavior.  That never should have happened, so I don't recommend following it as an example.

**Review by mattfenwick — CHANGES_REQUESTED** [review]
I'm not comfortable moving forward with an approach where a piece of code outside the test cases looks at the test name to determine which protocols to use.  That seems too magical and surprising for future maintainers.

File: test/e2e/network/netpol/model.go [hunk 13]
```diff
@@ -325,4 +355,7 @@ func (c *Container) Spec() v1.Container {
 			},
```
Comment (knabben) [inline test/e2e/network/netpol/model.go:355]: @rajaskakodkar you can use `Resources` as a direct field, and bring the comment up.

File: test/e2e/network/netpol/network_policy.go (hunk 8 above)
Comment (knabben) [reply test/e2e/network/netpol/network_policy.go:127]: for this test, we don't need to explicitly assert TCP, what about moving the test to the:

- UDP test suite.
- Create a TCP ingress rule.
- Assert UDP should not work.

otherwise we would need to recreted the models again passing TCP/UDP as protocols.

File: test/e2e/network/netpol/model.go (hunk 11 above)
Comment (jayunit100) [reply test/e2e/network/netpol/model.go:31]: How much smaller can you get then 1m?

**Comment by jayunit100** [comment]
Ah yeah clever hack rajas on the test names but maybe dumb it down a little :) just saw that .

File: test/e2e/network/netpol/network_policy.go (hunk 8 above)
Comment (jayunit100) [reply test/e2e/network/netpol/network_policy.go:127]: Ah yeah we can do that

File: test/e2e/network/netpol/model.go (hunk 12 above)
Comment (jayunit100) [reply test/e2e/network/netpol/model.go:33]: Hmmm you can do that in a constant? Not sure what the concern is here anyways ....

File: test/e2e/network/netpol/model.go (hunk 11 above)
Comment (jayunit100) [reply test/e2e/network/netpol/model.go:31]: for the analysis of the data, https://docs.google.com/spreadsheets/d/1a9ltOSV8AaGlKA1CZM_VW_D9lECc4W1x8hiyKuyLfoA/edit?usp=sharing

File: test/e2e/network/netpol/model.go (hunk 12 above)
Comment (jayunit100) [reply test/e2e/network/netpol/model.go:33]: ..... yeah , confirmed, i dont think you *can* call a function from a golang constant.... ?

File: test/e2e/network/netpol/model.go [hunk 14]
```diff
@@ -21,11 +21,28 @@ import (
...
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/kubernetes/test/e2e/framework"
 	imageutils "k8s.io/kubernetes/test/utils/image"
 )
 
+const (
+	CPURequest    = "1m"
+	MemoryRequest = "50Mi"
```
Comment (jayunit100) [inline test/e2e/network/netpol/model.go:32]: ran some experiments, imo this is a reasonable number but we could make it even smaller if we wanted to.  iirce k8s scheduler by default is in this region for binpacking  (100m)     
```
 docker container run -p 8000:8000 -d -t -i --memory-reservation=10m --name agn gcr.io/kubernetes-e2e-test-images/agnhost:2.6 net --serve 127.0.0.1:8000
```

**Comment by jayunit100** [comment]
@mattfenwick sounds in general like the main point is that tests should parameterize TCP + UDP directly ~ agree on that one.  but again kudos to rajas for the clever hack :).  

... rajas can you implement the more direct method matt mentioned for sending ports.... and then i think we can wrestle out the other details relatively easily....

File: test/e2e/network/netpol/model.go (hunk 12 above)
Comment (mattfenwick) [reply test/e2e/network/netpol/model.go:33]: You can use an `init` function if necessary.

File: test/e2e/network/netpol/model.go (hunk 11 above)
Comment (mattfenwick) [reply test/e2e/network/netpol/model.go:31]: I'm asking about memory request, cpu limit, memory limit 😄 -- does agnhost actually need that much?  I don't remember.

Is it possible to include a boiled-down summary of that data from the google spreadsheet as a comment?  Otherwise these just look like arbitrary constants with no context, and no way to tell if they're right or wrong.

Comment (jayunit100) [reply test/e2e/network/netpol/model.go:31]: +1 to include summary  as comments, and reference this issue (not as a proof that these numbers are perfect, but just as a "these empirically work and fence the total cpu usage")

File: test/e2e/network/netpol/model.go (hunk 12 above)
Comment (jayunit100) [reply test/e2e/network/netpol/model.go:33]: Is there an example elsewhere in k8s tests that you wanted rajas to follow ?

**Comment by jayunit100** [comment]
FWIW ( Am ok moving the code out of BeforeEach but IMO it’s a lateral move … ) just leaving that comment bc I assume we now might have a circular review where someone wonders why we no longer do beforeEach for pod setup :)

**Comment by mattfenwick** [comment]
Notes from syncup with @jayunit100 @rajaskakodkar : 

 - to get this PR merged, we'll back out [these changes](https://github.com/kubernetes/kubernetes/pull/104086#discussion_r685351088) and instead have each test specify which protocols it wants
 - future work (i.e. not part of this PR): instead of running containers on both ports 80 and 81 for all tests, let's just run containers on a single port for most tests, and multiple ports **ONLY** for the tests which specifically need that
 - future work (i.e. not part of this PR): let's reorganize tests so that they're grouped by the ports/protocols that they use

**Comment by aojea** [comment]
do we really need to play with the resource limits? or can we leave that part as last resort?

File: test/e2e/network/netpol/model.go (hunk 13 above)
Comment (rajaskakodkar) [reply test/e2e/network/netpol/model.go:355]: added this in the recent commit

**Comment by rajaskakodkar** [comment]
Summary of the analysis done to validate the experiments carried out in this PR:

We used a dataset[1] of 3 runs with the experimental changes and 3 runs without. The aim was to compare the time taken to run the tests and we have observed a reduction of ~2mins for almost all the tests.

The analysis can be viewed here https://docs.google.com/spreadsheets/d/1a9ltOSV8AaGlKA1CZM_VW_D9lECc4W1x8hiyKuyLfoA/edit?usp=sharing

[1] https://testgrid.k8s.io/sig-network-gce#presubmit-network-policies,%20google-gce&graph-metrics=test-duration-minutes&include-filter-by-regex=Netpol

Screenshot 2021-08-11 at 17-18-16 Netpol time analysis of experiments done in https github com kubernetes kubernetes pull 1
Screenshot 2021-08-11 at 17-17-34 Netpol time analysis of experiments done in https github com kubernetes kubernetes pull 1

**Comment by jayunit100** [comment]
I don't think backing the resource limits out is the end of the world necessarily.  

But in these tests given that the whole point of the pods is to be latent for the majority of their lifespan, why do we want to give them infinite CPU bandwidth?

**Comment by jayunit100** [comment]
@aojea (1) is a fair point.  me and @rajaskakodkar will collect more data and see ! :).    So that we can expedite this fix which will definetly reduce the flake rate, should we just get rid of the CPU/memory limits entirely ?   also lets squash those commits

**Comment by aojea** [comment]
I suggest starting changing only one thing (containers per pod) and
evaluate, we have time to do more changes later
If we add 2 changes at the same time we'll never known which one solved the
issue or in what proportion

File: test/e2e/network/netpol/model.go (hunk 12 above)
Comment (mattfenwick) [reply test/e2e/network/netpol/model.go:33]: https://golang.org/doc/effective_go#init

**Comment by jayunit100** [comment]
yeah we have consensus here.... lets update the PR and merge + iterate -> #104277 if there are any other improvements worth making.

**Comment by jayunit100** [comment]
... cc @squeed @astoycos can yall try to re-enable these tests post merge downstream .... ?

**Comment by rajaskakodkar** [comment]
@jayunit100 @aojea @mattfenwick I have updated the PR and squashed the commits PTAL

**Comment by jayunit100** [comment]
sure, looks like the same impl as before with just the requested refactors , lets see if the tests pass then on to https://github.com/kubernetes/kubernetes/issues/104277 !

**Comment by rajaskakodkar** [comment]
Running network-policies job again to check more data on how much time it takes or how flaky it is

/test pull-kubernetes-e2e-ubuntu-gce-network-policies

**Comment by aojea** [comment]
/kind bug
/sig network
/lgtm
/approve

The gain of moving from 3x containers per pod to 1 container per pod is a good improvement, we can iterate observing the corresponding testgrid after this merges
https://testgrid.k8s.io/sig-network-gce#network-policies,%20google-gce&width=90&graph-metrics=test-duration-minutes

Thanks for working on this
